File: lua/difference-of-squares/difference-of-squares.lua
-- difference-of-squares

local function ident(x) return x end
local function square(x) return x^2 end

local function sum(n, inner, outer)
    local sum = 0
    for i = 1, n do
        sum = sum + inner(i)
    end
    return outer(sum)
end

local function square_of_sum(n)  return sum(n, ident, square) end
local function sum_of_squares(n) return sum(n, square, ident) end

return {
    square_of_sum = square_of_sum,
    sum_of_squares = sum_of_squares,
    difference_of_squares = function(n) return square_of_sum(n) - sum_of_squares(n) end,
}

File: go/difference-of-squares/difference_of_squares.go
package diffsquares

func SquareOfSum(n int) int {
	var sum int
    /*
    for i := 1; i <= n; i++ {
        sum += i
    }
    */

    // sum: n(n+1)/2
    sum = n * (n + 1) / 2

	return sum * sum
}

func SumOfSquares(n int) int {
	var sum int
    /*
    for i := 1; i <= n; i++ {
        sum += i * i
    }
    */

    // sum of squares: n(n+1)(2n+1)/6
    sum = n * (n + 1) * (2 * n + 1) / 6

	return sum
}

func Difference(n int) int {
	return SquareOfSum(n) - SumOfSquares(n)
}

/* benchmarks
 *
 * using loops
 * BenchmarkSquareOfSum-2          15358201                67.61 ns/op            0 B/op          0 allocs/op
 * BenchmarkSumOfSquares-2         17112925                67.21 ns/op            0 B/op          0 allocs/op
 * BenchmarkDifference-2            8077932               147.1 ns/op             0 B/op          0 allocs/op
 *
 * using formulae
 * BenchmarkSquareOfSum-2          1000000000               1.084 ns/op           0 B/op          0 allocs/op
 * BenchmarkSumOfSquares-2         1000000000               0.5439 ns/op          0 B/op          0 allocs/op
 * BenchmarkDifference-2           1000000000               1.084 ns/op           0 B/op          0 allocs/op
 */
